File: apps/admin-web/src/api/volunteerManagement.ts
import { http } from "@/utils/http";

export type VolunteerServiceUnit = {
  id: string;
  unit_code: string;
  name: string;
  system_type: "CLASS_TEAM" | "GOVERNANCE" | "COMMITTEE_LINE" | "ACTIVITY";
  system_name: string;
  line_type: "LEARNING" | "OPERATIONS" | "DEVELOPMENT" | "GENERAL" | "SUPERVISION";
  line_name: string;
  parent_id?: string | null;
  parent_name?: string | null;
  home_shuku_org_unit_id?: string | null;
  home_shuku_name?: string | null;
  service_target_org_unit_id: string;
  service_target_name?: string | null;
  service_target_unit_type?: string | null;
  is_active: boolean;
  sort_order: number;
};

export type VolunteerPositionOption = {
  position_key: string;
  position_name: string;
  scope_level: string;
  capabilities: string[];
  capability_names: string[];
};

export type VolunteerAppointment = {
  id: number;
  member_id: number;
  member_name: string;
  member_status: string;
  home_shuku_name?: string | null;
  system_type?: string | null;
  system_name?: string | null;
  line_type?: string | null;
  line_name?: string | null;
  service_unit_id?: string | null;
  service_unit_name?: string | null;
  position_key: string;
  position_name: string;
  service_target_org_unit_id: string;
  service_target_name?: string | null;
  capabilities: string[];
  status: string;
  status_name: string;
  created_at?: string;
  ended_at?: string | null;
};

export type VolunteerRecommendationRule = {
  id: number;
  source_service_unit_id: string;
  source_service_unit_name: string;
  source_position_key: string;
  source_position_name: string;
  target_service_unit_id: string;
  target_service_unit_name: string;
  target_position_key: string;
  target_position_name: string;
  is_active: boolean;
};

const base = "/api/v1/volunteer-management";

export const getVolunteerServiceUnits = (params?: Record<string, unknown>) =>
  http.request<{ success: boolean; data: VolunteerServiceUnit[] }>("get", `${base}/service-units`, { params });

export const createVolunteerServiceUnit = (data: Omit<VolunteerServiceUnit, "id" | "system_name" | "line_name" | "parent_name" | "home_shuku_name" | "service_target_name" | "service_target_unit_type" | "is_active">) =>
  http.request<{ success: boolean; data: VolunteerServiceUnit }>("post", `${base}/service-units`, { data });

export const getVolunteerPositionOptions = (service_unit_id: string) =>
  http.request<{ success: boolean; data: { service_unit: VolunteerServiceUnit; positions: VolunteerPositionOption[] } }>(
    "get",
    `${base}/position-options`,
    { params: { service_unit_id } }
  );

export const getVolunteerAppointments = (params?: Record<string, unknown>) =>
  http.request<{ success: boolean; data: VolunteerAppointment[] }>("get", `${base}/appointments`, { params });

export const createVolunteerAppointment = (data: {
  member_id: number;
  service_unit_id: string;
  position_key: string;
  confirmation_note: string;
}) => http.request<{ success: boolean; data: { id: number; recommendations: VolunteerRecommendationRule[] } }>("post", `${base}/appointments`, { data });

export const changeVolunteerAppointmentStatus = (
  id: number,
  data: { status: "ACTIVE" | "SUSPENDED" | "ENDED" | "REVOKED"; reason: string }
) => http.request<{ success: boolean; data: { id: number; status: string } }>("post", `${base}/appointments/${id}/status`, { data });

export const getVolunteerRecommendationRules = (params?: Record<string, unknown>) =>
  http.request<{ success: boolean; data: VolunteerRecommendationRule[] }>("get", `${base}/recommendation-rules`, { params });

export const createVolunteerRecommendationRule = (data: {
  source_service_unit_id: string;
  source_position_key: string;
  target_service_unit_id: string;
  target_position_key: string;
}) => http.request<{ success: boolean; data: VolunteerRecommendationRule }>("post", `${base}/recommendation-rules`, { data });

export const acceptVolunteerRecommendation = (appointmentId: number, ruleId: number, confirmation_note: string) =>
  http.request<{ success: boolean; data: { target_appointment_id: number } }>(
    "post",
    `${base}/appointments/${appointmentId}/recommendations/${ruleId}/accept`,
    { data: { confirmation_note } }
  );

export const getVolunteerMigrationPreview = () =>
  http.request<{ success: boolean; data: { migration_executed: boolean; counts: Record<string, number>; entries: Array<Record<string, unknown>> } }>("get", `${base}/migration-preview`);

File: apps/platform-api/app/api/volunteer_management.py
"""HTTP surface for the dedicated Volunteer 2.0 appointment workspace."""

from __future__ import annotations

from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field, model_validator

from app.api.auth import require_permission
from app.services import volunteer_management as service


router = APIRouter(prefix="/api/v1/volunteer-management", tags=["volunteer-management"])


class ServiceUnitPayload(BaseModel):
    unit_code: str = Field(min_length=2, max_length=128)
    name: str = Field(min_length=1, max_length=255)
    system_type: Literal["CLASS_TEAM", "GOVERNANCE", "COMMITTEE_LINE", "ACTIVITY"]
    line_type: Literal["LEARNING", "OPERATIONS", "DEVELOPMENT", "GENERAL", "SUPERVISION"]
    parent_id: str | None = Field(default=None, max_length=64)
    home_shuku_org_unit_id: str | None = Field(default=None, max_length=64)
    service_target_org_unit_id: str = Field(min_length=1, max_length=64)
    sort_order: int = Field(default=0, ge=-100000, le=100000)


class AppointmentPayload(BaseModel):
    member_id: int = Field(gt=0)
    service_unit_id: str | None = Field(default=None, min_length=1, max_length=64)
    service_target_org_unit_id: str | None = Field(
        default=None, min_length=1, max_length=64
    )
    position_key: str = Field(min_length=3, max_length=64)
    confirmation_note: str = Field(default="学员管理页添加志工任职", max_length=1000)

    @model_validator(mode="after")
    def require_service_choice(self):
        if not self.service_unit_id and not self.service_target_org_unit_id:
            raise ValueError("请选择服务组织")
        return self


class AppointmentStatusPayload(BaseModel):
    status: Literal["ACTIVE", "SUSPENDED", "ENDED", "REVOKED"]
    reason: str = Field(default="学员管理页确认结束志工任职", max_length=1000)


class RecommendationRulePayload(BaseModel):
    source_service_unit_id: str = Field(min_length=1, max_length=64)
    source_position_key: str = Field(min_length=3, max_length=64)
    target_service_unit_id: str = Field(min_length=1, max_length=64)
    target_position_key: str = Field(min_length=3, max_length=64)


class RecommendationAcceptPayload(BaseModel):
    confirmation_note: str = Field(min_length=8, max_length=1000)


def _call(function, *args, **kwargs):
    try:
        return function(*args, **kwargs)
    except PermissionError as exc:
        raise HTTPException(403, str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(400, str(exc)) from exc


@router.get("/service-units")
def get_service_units(
    system_type: str | None = Query(default=None),
    line_type: str | None = Query(default=None),
    parent_id: str | None = Query(default=None, max_length=64),
    active_only: bool = Query(default=True),
    user: dict = Depends(require_permission("members:detail_view")),
) -> dict:
    return {
        "success": True,
        "data": _call(
            service.list_service_units,
            user["id"],
            system_type=system_type,
            line_type=line_type,
            parent_id=parent_id,
            active_only=active_only,
        ),
    }


@router.post("/service-units")
def add_service_unit(
    payload: ServiceUnitPayload,
    user: dict = Depends(require_permission("org:manage")),
) -> dict:
    return {"success": True, "data": _call(service.create_service_unit, user["id"], **payload.model_dump())}


@router.get("/position-options")
def get_position_options(
    service_unit_id: str = Query(min_length=1, max_length=64),
    user: dict = Depends(require_permission("members:detail_view")),
) -> dict:
    return {"success": True, "data": _call(service.list_position_options, user["id"], service_unit_id=service_unit_id)}


@router.get("/member-editor-catalog")
def get_member_editor_catalog(
    user: dict = Depends(require_permission("members:detail_view")),
) -> dict:
    return {"success": True, "data": _call(service.member_editor_catalog, user["id"])}


@router.get("/appointments")
def get_appointments(
    member_id: int | None = Query(default=None, gt=0),
    system_type: str | None = Query(default=None),
    line_type: str | None = Query(default=None),
    service_unit_id: str | None = Query(default=None, max_length=64),
    position_key: str | None = Query(default=None, max_length=64),
    home_shuku_org_unit_id: str | None = Query(default=None, max_length=64),
    service_target_org_unit_id: str | None = Query(default=None, max_length=64),
    member_name: str | None = Query(default=None, max_length=255),
    status: str | None = Query(default=None),
    user: dict = Depends(require_permission("members:detail_view")),
) -> dict:
    return {
        "success": True,
        "data": _call(
            service.list_appointments,
            user["id"],
            member_id=member_id,
            system_type=system_type,
            line_type=line_type,
            service_unit_id=service_unit_id,
            position_key=position_key,
            home_shuku_org_unit_id=home_shuku_org_unit_id,
            service_target_org_unit_id=service_target_org_unit_id,
            member_name=member_name,
            status=status,
        ),
    }


@router.post("/appointments")
def add_appointment(
    payload: AppointmentPayload,
    user: dict = Depends(require_permission("members:manage")),
) -> dict:
    return {"success": True, "data": _call(service.create_appointment, user["id"], **payload.model_dump())}


@router.post("/appointments/{appointment_id}/status")
def update_appointment_status(
    appointment_id: int,
    payload: AppointmentStatusPayload,
    user: dict = Depends(require_permission("members:manage")),
) -> dict:
    return {"success": True, "data": _call(service.change_appointment_status, user["id"], appointment_id, **payload.model_dump())}


@router.get("/recommendation-rules")
def get_recommendation_rules(
    source_service_unit_id: str | None = Query(default=None, max_length=64),
    source_position_key: str | None = Query(default=None, max_length=64),
    active_only: bool = Query(default=False),
    user: dict = Depends(require_permission("members:detail_view")),
) -> dict:
    return {
        "success": True,
        "data": _call(
            service.list_recommendation_rules,
            user["id"],
            source_service_unit_id=source_service_unit_id,
            source_position_key=source_position_key,
            active_only=active_only,
        ),
    }


@router.post("/recommendation-rules")
def add_recommendation_rule(
    payload: RecommendationRulePayload,
    user: dict = Depends(require_permission("org:manage")),
) -> dict:
    return {"success": True, "data": _call(service.create_recommendation_rule, user["id"], **payload.model_dump())}


@router.post("/appointments/{appointment_id}/recommendations/{rule_id}/accept")
def accept_recommended_appointment(
    appointment_id: int,
    rule_id: int,
    payload: RecommendationAcceptPayload,
    user: dict = Depends(require_permission("members:manage")),
) -> dict:
    return {
        "success": True,
        "data": _call(
            service.accept_recommendation,
            user["id"],
            source_appointment_id=appointment_id,
            recommendation_rule_id=rule_id,
            **payload.model_dump(),
        ),
    }


@router.get("/migration-preview")
def get_migration_preview(
    user: dict = Depends(require_permission("members:manage")),
) -> dict:
    return {"success": True, "data": _call(service.migration_preview, user["id"])}
